File: 00 - Orchestrator/orchestrator/credentials/oauth_report.py
"""Sanitized OAuth session reports for the Orchestrator."""

from __future__ import annotations

from datetime import UTC, datetime
from pathlib import Path
from typing import Any

from ..state import atomic_json_write


def utc_now_iso() -> str:
    return datetime.now(UTC).replace(microsecond=0).isoformat()


def write_oauth_report(state_dir: Path, report: dict[str, Any]) -> Path:
    target = Path(state_dir) / "oauth_latest_report.json"
    atomic_json_write(target, sanitize_report(report))
    return target


def sanitize_report(value: Any) -> Any:
    if isinstance(value, dict):
        output: dict[str, Any] = {}
        for key, child in value.items():
            lowered = str(key).lower()
            if lowered in {"access_token", "refresh_token", "authorization_code", "code", "id_token"}:
                output[key] = "[REDACTED]"
                continue
            output[key] = sanitize_report(child)
        return output
    if isinstance(value, list):
        return [sanitize_report(item) for item in value]
    return value

File: Client Frontend/client_frontend/credentials/oauth_report.js
import { mkdir } from "node:fs/promises";
import path from "node:path";

import { writeTextAtomically } from "../atomic_file.js";
import { OAUTH_REPORT_FILE } from "./types.js";

export function utcNowIso() {
  return new Date().toISOString();
}

function sanitizeReport(value) {
  if (Array.isArray(value)) {
    return value.map(sanitizeReport);
  }
  if (value && typeof value === "object") {
    return Object.fromEntries(
      Object.entries(value).map(([key, child]) => [
        key,
        ["access_token", "refresh_token", "authorization_code", "code", "id_token"].includes(key.toLowerCase())
          ? "[REDACTED]"
          : sanitizeReport(child)
      ])
    );
  }
  return value;
}

export async function writeOAuthReport(stateDir, report) {
  await mkdir(stateDir, { recursive: true });
  const target = path.join(stateDir, OAUTH_REPORT_FILE);
  await writeTextAtomically(target, `${JSON.stringify(sanitizeReport(report), null, 2)}\n`);
  return target;
}
